doc/neps/nep-0057-array-in-array-out.rst

```diff
@@ -0,0 +1,231 @@
+of nanoseconds. Consequently, some elementary arithmetic calculations will be
+slower. On the other hand, conversion from scalars to zero-rank arrays takes a
+few hundred nanoseconds, and many operations, such such as ufuncs and
+operations involving both scalars and rank-zero arrays require conversion from
+scalars to zero-rank arrays. These operations will be faster. We will not
+speculate as to whether this will have a net positive or net negative impact on
+user applications, but the net impact is expected to be small since impact on
+downstream library test suites has been minimal in testing.
+
+Backward compatibility
+----------------------
+
+The motivation of this proposal is to eliminate the surprises associated with
+a scalar result when a zero-rank array would be expected. However, existing
+code may rely on the current behavior, and this presents backward compatibility
+concerns.
+
+The main concern for user code is that the mutable zero-rank arrays that
+replace immutable scalars are no longer hashable. For instance, they cannot
+by used directly as keys of dictionaries, the argument of an ``lru_cache``
+-decorated function, etc. In all circumstances, tbe patch is simple: convert
+the zero-rank array to a scalar with empty-tuple indexing.
+
+Running the test suites of dependent libraries against a branch of NumPy that
+implements these changes has revealed a few other issues. <Library maintainers,
+please summaries these issues here.>
+
+Detailed description
+--------------------
+
+The new functionality will be used in much the same was as old functionality,
+except that sometimes authors will need to convert zero-rank arrays to scalars
+rather than converting scalars to zero-rank arrays. For instance, consider the
+following:
+
+.. code:: python
+
+   import numpy as np
+   rng = np.random.default_rng(85878653462722874072976519960992129768)
+   x = rng.standard_normal(size=10)
+   y = np.sum(x, axis=-1)
+   z = {y: 'a duck'}  # use scalar result as dictionary key
+   y = np.asarray(y)  # convert to array to allow mutation
+   y[y < 0] = np.nan
+
+The use of ``x`` as a dictionary key would need to become ``x[()]``, but ``z``
+no longer needs to be explicitly converted to an array.
+
+.. code:: python
+
+   z = {y[()]: 'a duck'}  # extract scalar to use as dictionary key
+   y[y < 0] = np.nan      # no conversion to array required
+
+Realistic examples can be found throughout the codebases of dependent
+libraries.
+
+Related work
+------------
+
+All known libraries that attempt to implement the array API standard
+(e.g. ``cupy``, ``torch``, ``jax.numpy``, ``dask.array``) return
+zero-rank arrays as specified by the standard.
+
+Implementation
+--------------
+
+To implement the NEP, the branch prepared by Sebastian needs to be merged,
+and dependent libraries will need to prepare releases that adapt to (and take
+advantage of) the changes. Branches of libraries including SciPy and Matplotlib
```

Comment:
@j-bowhay @pfackeldey @ksunden can you provide links to your branches?

Reply:
This is the link to Awkward Array with the NumPy branch provided by @seberg: https://github.com/scikit-hep/awkward/tree/pfackeldey/test_NEP57. There are only 2 failures and they are related to pandas. For Awkward Array the migration seems to be simple.

Reply:
Just catching up on notification now, happy to provide a branch for SciPy and read over this but will probably need a week or two to do so

Reply:
I have worked on it here: https://github.com/numpy/numpy/pull/29067 but there are still some issues around polynomials mostly I think, where the NumPy python code needs to properly deal with the changes (which is a bit tedious).

Reply:
It's ok to wait, @j-bowhay. We'll ping you when things are closer.